File: packages/core/src/sanitize.ts
import type {
  MatchedRule,
  SanitizeResult,
  SanitizerCatalog,
  SanitizerOptions,
  SanitizerRule
} from './types.js';

interface CompiledStripParam {
  kind: 'strip-param';
  source: SanitizerRule extends { source: infer S } ? S : never;
  provider: string;
  urlPattern: RegExp | null;
  paramPattern: RegExp;
  valuePattern: RegExp | null;
  exceptions: RegExp[];
  isReferralMarketing: boolean;
}

interface CompiledRawReplace {
  kind: 'raw-replace';
  source: CompiledStripParam['source'];
  provider: string;
  urlPattern: RegExp | null;
  pattern: RegExp;
  replacement: string;
  exceptions: RegExp[];
}

interface CompiledUnwrapRedirect {
  kind: 'unwrap-redirect';
  source: CompiledStripParam['source'];
  provider: string;
  urlPattern: RegExp | null;
  pattern: RegExp;
  captureGroup: number;
  matchPart: 'url' | 'pathname';
  targetEncoding: 'percent' | 'base64';
  prependScheme: 'http' | 'https' | undefined;
  targetTemplate: string | undefined;
  exceptions: RegExp[];
}

interface CompiledBlockDomain {
  kind: 'block-domain';
  source: CompiledStripParam['source'];
  provider: string;
  urlPattern: RegExp;
  exceptions: RegExp[];
}

type CompiledRule =
  | CompiledStripParam
  | CompiledRawReplace
  | CompiledUnwrapRedirect
  | CompiledBlockDomain;

function safeRegex(pattern: string, flags = ''): RegExp | null {
  try {
    return new RegExp(pattern, flags);
  } catch {
    return null;
  }
}

function compileExceptions(patterns: string[] | undefined): RegExp[] {
  if (!patterns) return [];
  const out: RegExp[] = [];
  for (const p of patterns) {
    const re = safeRegex(p, '');
    if (re) out.push(re);
  }
  return out;
}

function compileRule(rule: SanitizerRule): CompiledRule | null {
  switch (rule.kind) {
    case 'strip-param': {
      const paramPattern = safeRegex(`^(?:${rule.paramPattern})$`, 'i');
      if (!paramPattern) return null;
      const valuePattern = rule.valuePattern ? safeRegex(`^(?:${rule.valuePattern})$`, 'i') : null;
      if (rule.valuePattern && !valuePattern) return null;
      const urlPattern = rule.urlPattern ? safeRegex(rule.urlPattern, 'i') : null;
      if (rule.urlPattern && !urlPattern) return null;
      return {
        kind: 'strip-param',
        source: rule.source,
        provider: rule.provider,
        urlPattern,
        paramPattern,
        valuePattern,
        exceptions: compileExceptions(rule.exceptions),
        isReferralMarketing: rule.isReferralMarketing ?? false
      };
    }
    case 'raw-replace': {
      const pattern = safeRegex(rule.pattern, 'gi');
      if (!pattern) return null;
      const urlPattern = rule.urlPattern ? safeRegex(rule.urlPattern, 'i') : null;
      if (rule.urlPattern && !urlPattern) return null;
      return {
        kind: 'raw-replace',
        source: rule.source,
        provider: rule.provider,
        urlPattern,
        pattern,
        replacement: rule.replacement,
        exceptions: compileExceptions(rule.exceptions)
      };
    }
    case 'unwrap-redirect': {
      const pattern = safeRegex(rule.pattern, 'i');
      if (!pattern) return null;
      const urlPattern = rule.urlPattern ? safeRegex(rule.urlPattern, 'i') : null;
      if (rule.urlPattern && !urlPattern) return null;
      return {
        kind: 'unwrap-redirect',
        source: rule.source,
        provider: rule.provider,
        urlPattern,
        pattern,
        captureGroup: rule.captureGroup,
        matchPart: rule.matchPart ?? 'url',
        targetEncoding: rule.targetEncoding ?? 'percent',
        prependScheme: rule.prependScheme,
        targetTemplate: rule.targetTemplate,
        exceptions: compileExceptions(rule.exceptions)
      };
    }
    case 'block-domain': {
      const urlPattern = safeRegex(rule.urlPattern, 'i');
      if (!urlPattern) return null;
      return {
        kind: 'block-domain',
        source: rule.source,
        provider: rule.provider,
        urlPattern,
        exceptions: compileExceptions(rule.exceptions)
      };
    }
  }
}

function safeDecode(value: string): string {
  try {
    return decodeURIComponent(value);
  } catch {
    return value;
  }
}

function safeBase64Decode(value: string): string | null {
  try {
    const normalized = value.replace(/-/g, '+').replace(/_/g, '/');
    const padded = normalized.padEnd(normalized.length + ((4 - (normalized.length % 4)) % 4), '=');
    const binary = globalThis.atob(padded);
    const bytes = Uint8Array.from(binary, (char) => char.charCodeAt(0));
    return new TextDecoder().decode(bytes);
  } catch {
    return null;
  }
}

function templateTarget(template: string, match: RegExpExecArray): string {
  return template.replace(/\$([1-9])/g, (_, index: string) => match[Number(index)] ?? '');
}

function targetWithPrependedScheme(
  target: string,
  scheme: 'http' | 'https' | undefined
): string | null {
  if (!scheme) return target;
  try {
    new URL(target);
    return null;
  } catch {
    return `${scheme}://${target.replace(/^\/+/, '')}`;
  }
}

function matchesAny(url: string, patterns: RegExp[]): boolean {
  return patterns.some((re) => re.test(url));
}

function appliesTo(rule: CompiledRule, url: string): boolean {
  if (rule.kind !== 'block-domain' && rule.urlPattern && !rule.urlPattern.test(url)) {
    return false;
  }
  if (rule.kind === 'block-domain' && !rule.urlPattern.test(url)) return false;
  if (matchesAny(url, rule.exceptions)) return false;
  return true;
}

export type Sanitizer = (input: string) => SanitizeResult;

export function compileSanitizer(
  catalog: SanitizerCatalog,
  options: SanitizerOptions = {}
): Sanitizer {
  const {
    stripReferralMarketing = false,
    unwrapRedirects = true,
    domainBlocking = false
  } = options;

  const compiled: CompiledRule[] = [];
  for (const rule of catalog.rules) {
    const c = compileRule(rule);
    if (c) compiled.push(c);
  }

  return function sanitize(input: string): SanitizeResult {
    let parsed: URL;
    try {
      parsed = new URL(input);
    } catch {
      return { kind: 'unchanged', url: input };
    }

    let current = parsed.toString();
    const original = input;
    const strippedParams: string[] = [];
    const matchedRules: MatchedRule[] = [];

    for (const rule of compiled) {
      if (!appliesTo(rule, current)) continue;

      switch (rule.kind) {
        case 'block-domain': {
          if (!domainBlocking) continue;
          return {
            kind: 'blocked',
            original,
            via: { source: rule.source, provider: rule.provider, kind: rule.kind }
          };
        }

        case 'unwrap-redirect': {
          if (!unwrapRedirects) continue;
          let matchInput = current;
          if (rule.matchPart === 'pathname') {
            try {
              matchInput = new URL(current).pathname;
            } catch {
              continue;
            }
          }
          const m = rule.pattern.exec(matchInput);
          if (!m) continue;
          const captured = rule.targetTemplate
            ? templateTarget(rule.targetTemplate, m)
            : m[rule.captureGroup];
          if (typeof captured === 'string' && captured.length > 0) {
            const decodedTarget =
              rule.targetEncoding === 'base64'
                ? safeBase64Decode(safeDecode(captured))
                : safeDecode(captured);
            if (!decodedTarget) continue;
            const target = targetWithPrependedScheme(decodedTarget, rule.prependScheme);
            if (!target) continue;
            try {
              const targetUrl = new URL(target);
              return {
                kind: 'redirected',
                original,
                url: targetUrl.toString(),
                via: { source: rule.source, provider: rule.provider, kind: rule.kind }
              };
            } catch {
              // captured target isn't a valid URL — skip and continue
            }
          }
          continue;
        }

        case 'raw-replace': {
          const next = current.replace(rule.pattern, rule.replacement);
          if (next !== current) {
            current = next;
            matchedRules.push({
              source: rule.source,
              provider: rule.provider,
              kind: rule.kind
            });
          }
          continue;
        }

        case 'strip-param': {
          if (rule.isReferralMarketing && !stripReferralMarketing) continue;
          let urlObj: URL;
          try {
            urlObj = new URL(current);
          } catch {
            continue;
          }
          const search = stripFromQueryString(urlObj.search, rule.paramPattern, rule.valuePattern);
          const hash = stripFromQueryString(urlObj.hash, rule.paramPattern, rule.valuePattern);
          const removed = [...search.removed, ...hash.removed];
          if (removed.length === 0) continue;
          urlObj.search = search.result;
          urlObj.hash = hash.result;
          current = urlObj.toString();
          strippedParams.push(...removed);
          matchedRules.push({
            source: rule.source,
            provider: rule.provider,
            kind: rule.kind,
            detail: removed.join(',')
          });
          continue;
        }
      }
    }

    if (current === original) return { kind: 'unchanged', url: original };
    return {
      kind: 'cleaned',
      original,
      url: current,
      strippedParams,
      matchedRules
    };
  };
}

function stripFromQueryString(
  searchOrHash: string,
  paramPattern: RegExp,
  valuePattern: RegExp | null
): { result: string; removed: string[] } {
  if (!searchOrHash) return { result: searchOrHash, removed: [] };
  const prefix = searchOrHash.charAt(0);
  const body = prefix === '?' || prefix === '#' ? searchOrHash.slice(1) : searchOrHash;
  if (!body) return { result: searchOrHash, removed: [] };
  const pairs = body.split('&').filter(Boolean);
  const kept: string[] = [];
  const removed: string[] = [];
  for (const pair of pairs) {
    const eqIdx = pair.indexOf('=');
    const name = eqIdx === -1 ? pair : pair.slice(0, eqIdx);
    const value = eqIdx === -1 ? '' : pair.slice(eqIdx + 1);
    if (paramPattern.test(name) && (!valuePattern || valuePattern.test(value))) {
      removed.push(name);
    } else {
      kept.push(pair);
    }
  }
  if (removed.length === 0) return { result: searchOrHash, removed: [] };
  if (kept.length === 0) return { result: '', removed };
  return { result: prefix + kept.join('&'), removed };
}

File: crates/url-sanitize-core/src/sanitize.rs
use base64::{engine::general_purpose, Engine as _};
use percent_encoding::percent_decode_str;
use url::Url;

use crate::compile::{CompiledBody, CompiledRule};
use crate::types::{
    MatchedRule, RedirectPrependScheme, RedirectTargetEncoding, SanitizeResult, SanitizerOptions,
};

fn applies_to(rule: &CompiledRule, current: &str) -> bool {
    if let Some(re) = &rule.url_pattern {
        if !re.is_match(current) {
            return false;
        }
    }
    for ex in &rule.exceptions {
        if ex.is_match(current) {
            return false;
        }
    }
    true
}

fn safe_decode(s: &str) -> String {
    percent_decode_str(s).decode_utf8_lossy().into_owned()
}

fn safe_base64_decode(s: &str) -> Option<String> {
    let normalized = s.replace('-', "+").replace('_', "/");
    let padded = {
        let remainder = normalized.len() % 4;
        if remainder == 0 {
            normalized.clone()
        } else {
            format!("{}{}", normalized, "=".repeat(4 - remainder))
        }
    };

    for engine in [
        &general_purpose::STANDARD,
        &general_purpose::URL_SAFE,
        &general_purpose::URL_SAFE_NO_PAD,
    ] {
        if let Ok(bytes) = engine.decode(&padded) {
            return Some(String::from_utf8_lossy(&bytes).into_owned());
        }
    }
    None
}

fn template_target(template: &str, caps: &regex_lite::Captures<'_>) -> String {
    let mut out = String::new();
    let mut chars = template.chars().peekable();
    while let Some(ch) = chars.next() {
        if ch == '$' {
            if let Some(next) = chars.peek().copied() {
                if ('1'..='9').contains(&next) {
                    chars.next();
                    let index = next.to_digit(10).unwrap_or(0) as usize;
                    if let Some(mat) = caps.get(index) {
                        out.push_str(mat.as_str());
                    }
                    continue;
                }
            }
        }
        out.push(ch);
    }
    out
}

fn target_with_prepended_scheme(
    target: &str,
    scheme: Option<RedirectPrependScheme>,
) -> Option<String> {
    let Some(scheme) = scheme else {
        return Some(target.to_string());
    };
    if Url::parse(target).is_ok() {
        return None;
    }
    Some(format!(
        "{}://{}",
        scheme.as_str(),
        target.trim_start_matches('/')
    ))
}

/// Strip params matching `pattern` from a raw query/fragment body (no leading
/// `?`/`#`). Returns (kept body, removed names). Pair format preserved verbatim
/// for kept entries.
fn strip_param_pairs(
    body: &str,
    pattern: &regex_lite::Regex,
    value_pattern: Option<&regex_lite::Regex>,
) -> (String, Vec<String>) {
    let mut kept: Vec<&str> = Vec::new();
    let mut removed: Vec<String> = Vec::new();
    for pair in body.split('&').filter(|p| !p.is_empty()) {
        let name = match pair.find('=') {
            Some(idx) => &pair[..idx],
            None => pair,
        };
        let value = match pair.find('=') {
            Some(idx) => &pair[idx + 1..],
            None => "",
        };
        if pattern.is_match(name) && value_pattern.map_or(true, |p| p.is_match(value)) {
            removed.push(name.to_string());
        } else {
            kept.push(pair);
        }
    }
    (kept.join("&"), removed)
}

pub(crate) fn sanitize_with(
    rules: &[CompiledRule],
    options: &SanitizerOptions,
    input: &str,
) -> SanitizeResult {
    let parsed = match Url::parse(input) {
        Ok(u) => u,
        Err(_) => {
            return SanitizeResult::Unchanged {
                url: input.to_string(),
            }
        }
    };

    let original = input.to_string();
    let mut current = parsed.to_string();
    let mut stripped_params: Vec<String> = Vec::new();
    let mut matched_rules: Vec<MatchedRule> = Vec::new();

    for rule in rules {
        if !applies_to(rule, &current) {
            continue;
        }
        match &rule.body {
            CompiledBody::BlockDomain => {
                if !options.domain_blocking {
                    continue;
                }
                return SanitizeResult::Blocked {
                    original,
                    via: MatchedRule {
                        source: rule.source,
                        provider: rule.provider.clone(),
                        kind: rule.kind,
                        detail: None,
                    },
                };
            }
            CompiledBody::UnwrapRedirect {
                pattern,
                capture_group,
                match_part,
                target_encoding,
                prepend_scheme,
                target_template,
            } => {
                if !options.unwrap_redirects() {
                    continue;
                }
                let match_input = if matches!(match_part, crate::types::RedirectMatchPart::Pathname)
                {
                    let Ok(url_obj) = Url::parse(&current) else {
                        continue;
                    };
                    url_obj.path().to_string()
                } else {
                    current.clone()
                };
                let Some(caps) = pattern.captures(&match_input) else {
                    continue;
                };
                let captured = if let Some(template) = target_template {
                    template_target(template, &caps)
                } else {
                    let Some(mat) = caps.get(*capture_group) else {
                        continue;
                    };
                    mat.as_str().to_string()
                };
                if captured.is_empty() {
                    continue;
                }
                let decoded = match target_encoding {
                    RedirectTargetEncoding::Percent => safe_decode(&captured),
                    RedirectTargetEncoding::Base64 => {
                        let percent_decoded = safe_decode(&captured);
                        let Some(decoded) = safe_base64_decode(&percent_decoded) else {
                            continue;
                        };
                        decoded
                    }
                };
                let Some(target) = target_with_prepended_scheme(&decoded, *prepend_scheme) else {
                    continue;
                };
                let Ok(target_url) = Url::parse(&target) else {
                    continue;
                };
                return SanitizeResult::Redirected {
                    original,
                    url: target_url.to_string(),
                    via: MatchedRule {
                        source: rule.source,
                        provider: rule.provider.clone(),
                        kind: rule.kind,
                        detail: None,
                    },
                };
            }
            CompiledBody::RawReplace {
                pattern,
                replacement,
            } => {
                let next = pattern.replace_all(&current, replacement.as_str());
                if next != current {
                    current = next.into_owned();
                    matched_rules.push(MatchedRule {
                        source: rule.source,
                        provider: rule.provider.clone(),
                        kind: rule.kind,
                        detail: None,
                    });
                }
            }
            CompiledBody::StripParam {
                param_pattern,
                value_pattern,
                is_referral_marketing,
            } => {
                if *is_referral_marketing && !options.strip_referral_marketing {
                    continue;
                }
                let Ok(mut url_obj) = Url::parse(&current) else {
                    continue;
                };

                let mut all_removed: Vec<String> = Vec::new();

                // Query
                if let Some(q) = url_obj.query() {
                    let (kept, removed) =
                        strip_param_pairs(q, param_pattern, value_pattern.as_ref());
                    if !removed.is_empty() {
                        if kept.is_empty() {
                            url_obj.set_query(None);
                        } else {
                            url_obj.set_query(Some(&kept));
                        }
                        all_removed.extend(removed);
                    }
                }

                // Fragment-as-query
                if let Some(f) = url_obj.fragment() {
                    let (kept, removed) =
                        strip_param_pairs(f, param_pattern, value_pattern.as_ref());
                    if !removed.is_empty() {
                        if kept.is_empty() {
                            url_obj.set_fragment(None);
                        } else {
                            url_obj.set_fragment(Some(&kept));
                        }
                        all_removed.extend(removed);
                    }
                }

                if all_removed.is_empty() {
                    continue;
                }
                current = url_obj.to_string();
                let detail = all_removed.join(",");
                stripped_params.extend(all_removed);
                matched_rules.push(MatchedRule {
                    source: rule.source,
                    provider: rule.provider.clone(),
                    kind: rule.kind,
                    detail: Some(detail),
                });
            }
        }
    }

    if current == original {
        return SanitizeResult::Unchanged { url: original };
    }
    SanitizeResult::Cleaned {
        original,
        url: current,
        stripped_params,
        matched_rules,
    }
}
